lib/agent.js

```diff
@@ -0,0 +1,79 @@
+'use strict'
+
+const { InvalidArgumentError, InvalidReturnValueError, NotSupportedError } = require('./core/errors')
+const Pool = require('./pool')
+const { getPath, weakCache, checkForWeakRefAndFinalizationRegistrySupport } = require('./core/util')
+
+const kAgentFactory = Symbol('agent factory')
+
+class Agent {
+  constructor (opts) {
+    /* istanbul ignore next */
+    checkForWeakRefAndFinalizationRegistrySupport()
+
+    this[kAgentFactory] = weakCache(origin => new Pool(origin, opts))
+  }
+
+  get (origin) {
+    if (typeof origin !== 'string' || origin === '') {
+      throw new InvalidArgumentError('Origin must be a non-empty string.')
+    }
+
+    return this[kAgentFactory](origin)
+  }
+}
+
+let globalAgent
+try {
+  globalAgent = new Agent()
+} catch (err) {
+  if (err instanceof NotSupportedError) {
+    // silently fail to set globalAgent due to unsupported environment
+    globalAgent = null
+  }
+}
+
+function setGlobalAgent (agent) {
+  /* istanbul ignore next */
+  checkForWeakRefAndFinalizationRegistrySupport()
+
+  if (typeof agent.get === 'function') {
+    globalAgent = agent
+  } else {
+    throw new InvalidArgumentError('Argument agent must implement Agent')
+  }
+}
+
+function dispatchFromAgent (requestType) {
+  return (url, { agent = globalAgent, method = 'GET', ...opts } = {}, ...additionalArgs) => {
+    /* istanbul ignore next */
+    checkForWeakRefAndFinalizationRegistrySupport()
+
+    if (url === undefined || url === null) {
+      throw new InvalidArgumentError('Argument url must be defined')
+    }
+
+    // if url is a string transform into URL inst. otherwise use as is
+    url = typeof url === 'string' ? new URL(url) : url
```

Comment:
The last thing I am bringing up because I have seen it brought up in the past, using `URL` here is acceptable in terms of performance drop?

Reply:
This api is more for convenience than performance so I don't see it as a huge issue. But yea, there is room for optimization.